File: scripts/generateCnvReport.R
library(GenomicRanges, quietly=T)
library(VariantAnnotation, quietly=T)
library(stringr, quietly=T)

# Author: Christopher Medway
# Use: RScript generateCnvReport.R <seqId> <panel>
# Description: compiles a single run-wide report of CNVs, merging CNV calls
# from MANTA and ExomeDepth

# input arguments
roi <- commandArgs(trailingOnly = T)[1] #The bed file containing the roi
exomeDepthMatrics <- commandArgs(trailingOnly = T)[2]#Path to the exome metrics file
high_cov_bams <- commandArgs(trailingOnly = T)[3] #Path to the high coverage bam file
depth_folder <- commandArgs(trailingOnly = T)[4] #Path to the directory containing the depth summary files
exome_depth_dir <- commandArgs(trailingOnly = T)[5] #The folder containing the exome_depth results
manta_dir <- commandArgs(trailingOnly = T)[6] #The folder containing the manta results
out_name <- commandArgs(trailingOnly = T)[7] #Where to put the output csv


# load exome depth QC (r2 values)
exomeDepthMatricsDf <- read.table(exomeDepthMatrics, stringsAsFactors = F, header = T)

print (high_cov_bams)


# get list of samples that have CNV data
sampleId <- unlist(lapply(stringr::str_split(read.table(high_cov_bams, stringsAsFactors=F)[,1], pattern="/"), function(x) x[3]))
print(sampleId)
sampleId <-  unlist(lapply(stringr::str_split(sampleId, pattern="_"), function(x) paste0(x[1],"_", x[2])))
print (sampleId)


# generate initial QC flags where all samples = PASS
passQC <- data.frame(sampleId, "QC" = "PASS", stringsAsFactors = F)

# flag any sample with R@ < 0.98 as a FAIL
if (sum(exomeDepthMatricsDf$Correlation < 0.98) > 0) {
    fails <- sub(exomeDepthMatricsDf[exomeDepthMatricsDf$Correlation < 0.98,"BamPath"], perl = T, pattern = ".+_(\\w+).bam", replacement = "\\1")
    passQC[passQC$sampleId %in% fails, "QC"] <- "R2<0.98"
}

# add average depth for each sample
for (s in sampleId) {
	

    exomeDepthCoverage <- paste0(depth_folder,s,"_DepthOfCoverage.sample_summary")
    exomeDepthCoverageDf <- read.table(exomeDepthCoverage, stringsAsFactors = F, header = T, fill = T)
    passQC[passQC$sampleId %in% s, "Depth"] <- exomeDepthCoverageDf$mean[1]
}

# main function - loop over sample and compile MANTA / ExoneDepth data
mergedDf <- lapply(sampleId, function(sample) {
    
    qc    <- passQC[passQC$sampleId %in% sample,"QC"]
    depth <- passQC[passQC$sampleId %in% sample, "Depth"] 
    
    exomeDepth <- paste0(exome_depth_dir, sample, "_final_cnv_fixed.vcf.gz")
    manta      <- paste0(manta_dir, sample,"_diploidSV.vcf.gz")
    
    ## extract information from exomeDepth
    vcf    <- VariantAnnotation::readVcf(exomeDepth, "hg19")
    
    if (dim(vcf)[1] > 0) {
        chr     <- as.vector(vcf@rowRanges@seqnames)
        start   <- as.numeric(vcf@rowRanges@ranges@start)
        end     <- as.numeric(vcf@info$END)
        ref     <- as.character(vcf@fixed$REF)
        alt     <- as.character(vcf@fixed$ALT@unlistData)
        type    <- vcf@fixed$ALT@unlistData
        regions <- vcf@info$Regions
        
        if (any(is.na(end))) {end[is.na(end)] <- (start[is.na(end)] + 1)}
        
    } else {
        
        chr     <- NA
        start   <- NA
        end     <- NA
        ref     <- NA
        alt     <- NA
        type    <- NA
        regions <- NA
    }
    
    exomeDepthDF <- data.frame(
        sample,
        "method" = "exomeDepth",
        chr,
        start,
        end,
        ref,
        type,
        regions,
        qc,
        depth,
        stringsAsFactors = F
    )
    
    rm(vcf, chr, start, end, ref, alt, regions, type)
    

    
    ## extract information from MANTA
    vcf     <- VariantAnnotation::readVcf(manta, "hg19")
    
    if (dim(vcf)[1] > 0) {
        chr     <- as.vector(vcf@rowRanges@seqnames)
        start   <- as.numeric(vcf@rowRanges@ranges@start)
        end     <- as.numeric(vcf@info$END)
        ref     <- as.character(vcf@fixed$REF)
        alt     <- as.character(vcf@fixed$ALT@unlistData)
        type    <- vcf@info$SVTYPE
        
        if (any(is.na(end))) {end[is.na(end)] <- start[is.na(end)] + 1}
        
    } else {
        
        chr   <- NA
        start <- NA
        end   <- NA
        ref   <- NA
        alt   <- NA
        type  <- NA
        }
    
    
    mantaDF <- data.frame(
        sample,
        "method" = "manta",
        chr,
        start,
        end,
        ref,
        type,
        "regions" = "-",
        qc = "PASS",
        depth,
        stringsAsFactors = F
    )
    
    rm(vcf, chr, start, end, ref, alt, type)
    
    df <- rbind(exomeDepthDF, mantaDF)
    dfNonNull <- df[!is.na(df$chr),]
    dfNull    <- df[is.na(df$chr),]
    
    ## annotate with gene names
    roiDf <- read.table(roi, stringsAsFactors = F)
    roiDf$gene <- stringr::str_extract(string = roiDf$V4, pattern="(\\w+)")
    roiUnified <- do.call(rbind, lapply(split(roiDf, roiDf$gene), function(x) {data.frame("chr" = x$V1[1], "start" = min(x$V2), "end" = max(x$V3), "gene" = x$gene[1])}))

    roiGR <- GenomicRanges::GRanges(seqnames = roiUnified$chr, ranges = IRanges(start = roiUnified$start, end = roiUnified$end))
    
    gr <- GenomicRanges::GRanges(seqnames = dfNonNull$chr, ranges = IRanges(start = as.numeric(dfNonNull$start), end = as.numeric(dfNonNull$end)))
    ol <- findOverlaps(gr, roiGR)
    dfNonNull[queryHits(ol),"gene"] <- roiUnified[subjectHits(ol),"gene"]
    dfNonNull <- dfNonNull[order(as.numeric(dfNonNull$chr), as.numeric(dfNonNull$start)),]
    
    if (dim(dfNull)[1] == 0) {
        return(dfNonNull)
    } else {
        dfNull$gene <- NA
        return(rbind(dfNonNull, dfNull))
    }
})

mergedDf <- do.call(rbind, mergedDf)

mergedDf[mergedDf$depth < 160 & mergedDf$qc == "R2<0.98","qc"] <- "R2<0.98;Depth<160"
mergedDf[mergedDf$depth < 160 & mergedDf$qc == "PASS","qc"]    <- "Depth<160"

# write to current directory (run folder)
write.table(mergedDf, file = paste0(out_name), row.names = F, sep = ",", quote = F)

File: scripts/generateCNVReport.py
import pandas
import allel
import argparse
from pathlib import Path

# Get the script arguments
parser = argparse.ArgumentParser()
parser.add_argument('--runid')
parser.add_argument('--output')
parser.add_argument('--bed')
parser.add_argument('--exome_metrics')
parser.add_argument('--manta_dir')
parser.add_argument('--exome_dir')
parser.add_argument('--coverage_dir')

args = parser.parse_args()
runid = args.runid
output_file = args.output
bed = args.bed
manta_dir = args.manta_dir
exome_dir = args.exome_dir
coverage_dir = args.coverage_dir
exome_metrics = args.exome_metrics

# Utility Functions
def fix_end_position(df):
	"""
	Fixes the END columns:

	If the END column is equal to -1 then change the end columns \
	to the start position + 1

	"""

	if df['END'] != 'NA':

		if int(df['END']) == -1:

			return int(df['POS']) + 1

	return df['END']

def assign_qc(df):
	"""
	Assign a QC status e.g. is r squared is less than a certain value

	"""

	qc_comment = []

	if df['Correlation'] < 0.98:

		qc_comment.append('R2<0.98')

	if df['depth'] < 160:

		qc_comment.append('Depth<160')

	if len(qc_comment) == 0:

		return 'PASS'

	else:

		return '|'.join(qc_comment)

def get_overlapping_genes(df, grouped_bed):
	"""
	Get the custom roi file and check to see if any of our SVs overlap with the regions.

	"""

	sv_start = df['POS']

	if sv_start == 'NA':

		return 'NA'

	sv_chrom = str(df['CHROM'])
	sv_start = int(df['POS'])
	sv_end = int(df['END'])

	gene_list = []

	for gene in grouped_bed.itertuples():

		gene_chrom = str(gene.Chrom)
		gene_start = int(gene.Start)
		gene_end = int(gene.End)

		# variant start is within the gene
		if gene_chrom == sv_chrom and gene_start <= sv_start <= gene_end:

			gene_list.append(gene.gene)

		# variant end is within the gene
		elif gene_chrom == sv_chrom and gene_start <= sv_end <= gene_end:

			gene_list.append(gene.gene)

		# variant start is before gene and variant end is after gene e.g. variant engulfs gene.
		elif gene_chrom == sv_chrom and (sv_start <= gene_start and sv_end >= gene_end):

			gene_list.append(gene.gene)

	if len(gene_list) > 0:

		return '|'.join(list(set(gene_list)))

	else:

		return 'NA'


# Create main dataframe
dataFrame = pandas.DataFrame()

# Read exome depth metrics file to get high coverage samples
exome_depth_metrics = pandas.read_table(exome_metrics) 
num_rows_exome_depth_metrics = exome_depth_metrics.shape[0]

# Add sampleid column to exome depth
exome_depth_metrics['sampleid'] = exome_depth_metrics['BamPath'].apply(lambda x: Path(x).stem.replace('_final', ''))

samples_list = list(exome_depth_metrics.groupby('sampleid').count().index)


# Read BED file
bed_file = pandas.read_csv(bed, names=['Chrom', 'Start', 'End', 'Comment'], sep='\t', header=None)

# Extract gene to new column
bed_file['gene'] = bed_file['Comment'].apply(lambda x: x.split('.')[0])

# Groupby gene and get the smallest start and largest end for each gene.
grouped_bed = bed_file.groupby(['Chrom', 'gene']).agg({'Start': 'min', 'End': 'max'})

grouped_bed = grouped_bed.reset_index()


#Loop through all samples and read in the relevant files
for sample in samples_list:

	sample_name = sample.split('_')[4]
	sample_number = sample.split('_')[5]
	
	# Get the mean depth from the summary file
	depth_of_coverage_summary = pandas.read_table(sample_name + '/' +  sample + '_DepthOfCoverage.sample_summary')
	depth_of_coverage_summary_mean = depth_of_coverage_summary['mean'][1]
	
	#Create dataframe from manta vcf
	manta_df = allel.vcf_to_dataframe(manta_dir + '/' + sample_name + '_' + sample_number + '_diploidSV.vcf.gz' , fields=['*'])

	if manta_df is not None:

		manta_df_2 = manta_df[['CHROM', 'POS', 'END', 'REF', 'SVTYPE']]
		manta_df_2.columns = ['CHROM', 'POS', 'END', 'REF', 'SVTYPE']
		manta_df_2['Regions'] = '-'
		manta_df_2['QC'] = 'PASS'
		manta_df_2['method'] = 'Manta'
		manta_df_2['ALT_1'] = manta_df_2['SVTYPE']

	else:


		# Make a dataframe with a single row with everything set to NA
		data = [{'CHROM':'NA', 'POS':'NA', 'END':'NA', 'REF':'NA', 'SVTYPE':'NA', 'Regions':'-'}]
		manta_df_2 = pandas.DataFrame(data)
		manta_df_2 = manta_df_2[['CHROM', 'POS', 'END', 'REF', 'SVTYPE', 'Regions']]
		manta_df_2['QC'] = 'PASS'
		manta_df_2['method'] = 'Manta'
		manta_df_2['ALT_1'] = manta_df_2['SVTYPE']

	
	#Create dataframe from Exome depth vcf
	exome_df = allel.vcf_to_dataframe(exome_dir + '/' + sample + '_final_cnv_fixed.vcf.gz' , fields=['*'])

	if exome_df is not None:

		exome_df_2 = exome_df[['CHROM', 'POS', 'END', 'REF', 'ALT_1', 'Regions']]
		exome_df_2['QC'] = 'PASS'
		exome_df_2['method'] = 'exomeDepth'

	else:
		exome_data = [{'CHROM':'NA', 'POS':'NA', 'END':'NA', 'REF':'NA', 'ALT_1':'NA', 'Regions':'NA'}]
		exome_df_2 = pandas.DataFrame(exome_data)
		exome_df_2 = exome_df_2[['CHROM', 'POS', 'END', 'REF', 'ALT_1', 'Regions']]
		exome_df_2['QC'] = 'PASS'
		exome_df_2['method'] = 'exomeDepth'


	#Combine both Manta and ExomeDepth dataframes and add depth and sample ID
	if (exome_df_2 is not None) and (manta_df_2 is not None):

		manta_exome= exome_df_2.append(manta_df_2)
		manta_exome['depth'] = depth_of_coverage_summary_mean
		manta_exome['sampleid'] = sample

	elif (exome_df_2 is None) and (manta_df_2 is not None):

		manta_exome= manta_df_2
		manta_exome['depth']= depth_of_coverage_summary_mean
		manta_exome['sampleid']= sample

	elif (exome_df_2 is not None) and (manta_df_2 is None):

		manta_exome= exome_df_2
		manta_exome['depth'] = depth_of_coverage_summary_mean
		manta_exome['sampleid'] = sample

	else:

		manta_exome = None

	if manta_exome is not None:

		manta_exome['END'] = manta_exome.apply(fix_end_position,axis=1)

		# Annotate with data from the exome depth logs
		manta_exome = pandas.merge(manta_exome, exome_depth_metrics, left_on = 'sampleid', right_on = 'sampleid')
		manta_exome['QC'] = manta_exome.apply(assign_qc, axis=1)

		# Add gene annotation
		manta_exome['Gene'] = manta_exome.apply(get_overlapping_genes, axis=1, args=(grouped_bed,))

		#Add sample dataframe to main dataframe containing all samples
		dataFrame = dataFrame.append(manta_exome)

#output the dataframe containing the CNVs for all sample as a TSV
if manta_exome is not None:

	final_df = dataFrame[['sampleid', 'method', 'CHROM', 'POS', 'END', 'REF', 'ALT_1', 'Regions', 'QC', 'depth', 'Gene']]

	# Remove < and > from <DUP> and <DEL>
	final_df['ALT_1'] = final_df['ALT_1'].apply(lambda x: x.strip('<').strip('>'))

	# Create new columns with different names
	final_df['Sample'] = final_df['sampleid']
	final_df['Method'] = final_df['method']
	final_df['Chr'] = final_df['CHROM']
	final_df['Start'] = final_df['POS']
	final_df['End'] = final_df['END']
	final_df['Ref'] = final_df['REF']
	final_df['Type'] = final_df['ALT_1']
	final_df['Depth'] = final_df['depth']

	# Write to TSV
	final_df[['Sample', 'Method', 'Chr', 'Start', 'End', 'Ref', 'Type', 'Regions', 'QC', 'Depth', 'Gene']].to_csv(output_file, index=False, sep=',')













